File: lib/cfr/builtins/libs/dart/ui/imageShader.dart
import 'dart:core';
import 'dart:typed_data';
import 'dart:ui';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedImageShader extends VMManagedBox<ImageShader> {
  VMManagedImageShader(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.toString(),
      ];
    });
    table['getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hashCode,
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final ImageShader vmObject;
}

class RTManagedImageShader extends ImageShader implements Box<ImageShader> {
  RTManagedImageShader(
      Image image, TileMode tmx, TileMode tmy, Float64List matrix4,
      {required this.table, required this.hydroState})
      : super(
          image,
          tmx,
          tmy,
          matrix4,
        ) {
    table['vmObject'] = vmObject;
    table['unwrap'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [unwrap()];
    });
    table['_dart_toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [super.toString()];
    });
    table['_dart_getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [super.hashCode];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  ImageShader unwrap() => this;
  ImageShader get vmObject => this;
  @override
  String toString() {
    Closure closure = table["__tostring"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }

  @override
  int get hashCode {
    Closure closure = table["getHashCode"];
    return closure.dispatch([table], parentState: hydroState)[0];
  }
}

void loadImageShader(
    {required HydroState hydroState, required HydroTable table}) {
  table['imageShader'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      RTManagedImageShader(
          maybeUnBoxAndBuildArgument<Image, dynamic>(luaCallerArguments[1],
              parentState: hydroState),
          maybeUnBoxEnum(
              values: TileMode.values, boxedEnum: luaCallerArguments[2]),
          maybeUnBoxEnum(
              values: TileMode.values, boxedEnum: luaCallerArguments[3]),
          maybeUnBoxAndBuildArgument<Float64List, dynamic>(
              luaCallerArguments[4],
              parentState: hydroState),
          table: luaCallerArguments[0],
          hydroState: hydroState)
    ];
  });
  registerBoxer<ImageShader>(boxer: (
      {required ImageShader vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedImageShader(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}

File: runtime/dart/ui/imageShader.ts
import { IFloat64List } from "../typed_data/float64List";
import { IImage } from "./image";
import { IShader } from "./shader";
import { TileMode } from "./tileMode";
declare const dart: {
    ui: {
        imageShader: (
            this: void,
            imageShader: IImageShader,
            image: IImage,
            tmx: TileMode,
            tmy: TileMode,
            matrix4: IFloat64List
        ) => IImageShader;
    };
};
export interface IImageShader {
    toString: () => string;
    getHashCode: () => number;
}
export class ImageShader implements IShader {
    public constructor(
        image: IImage,
        tmx: TileMode,
        tmy: TileMode,
        matrix4: IFloat64List
    ) {
        dart.ui.imageShader(this, image, tmx, tmy, matrix4);
    }
    private readonly _dart_toString: () => string = undefined as any;
    private readonly _dart_getHashCode: () => number = undefined as any;
    public toString(): string {
        return this._dart_toString();
    }
    public getHashCode(): number {
        return this._dart_getHashCode();
    }
}
